text/3371-cargo-target-base-dir.md

```diff
@@ -0,0 +1,402 @@
+
+It won't work in the `config.toml` but it will work with the environment variable and the command line option, both of which override the TOML config.
+
+It is certainly possible to add at least `{home}` and `{cargo_home}` but it can be done in the future without interfering at all with `{manifest-path-hash}`, making it relatively easy to defer and add if there is demand.
+
+## Use templated `CARGO_TARGET_DIR` as the default instead of `target`
+
+This option has several complications I'm not sure how to resolve:
+
+1. How do we decide on good platform defaults ?
+  - Subsequently, when platform defaults are decided, how do we ensure a new platform has a good default too ?
+  - `CARGO_HOME` is already criticized for being both a *cache* and a *config* home (using the XDG spec semantics), adding more local cache to it in the form of `CARGO_HOME/target-base-dir/` would not improve the situation and should probably not be done, but, if no good alternatives are found, there is precedent to use it for this.
+2. How do we communicate on said default values ?
+3. This would probably break backward compatibility and lots of tools ? We could heavily advertise the option in the Rust book and Cargo's documentation but making it the default is probably not something we will be able (or even willing) to do any time soon. Note that having forward links active by default (see relevant section earlier in the RFC) will help offset a lot of the problems here.
+
+### We really want to do this, how do we do it ?
+
+Well, first, advertising of the option and its behaviour, as well as the forward link behaviour (so people know we're not just breaking tools for fun). After that, it becomes necessary to test it quite heavily to really ensure nothing has broken irremediably.
+
+1. Introduce it as the default behaviour in nightly, wait for a few stable releases so that beta has it too and it can start being used for a few months after that so that esoteric setups can also try it.
+2. Write a post saying "we'll do the change in version X" (where current version is like X-2 to warn 3 months before at least ?) and then only apply the change to projects where there is no untemplated `CARGO_TARGET_DIR` config set.
+
+### A user still wants the current behaviour and not this RFC's
+
+- We could use a config option to entirely deactivate it
+- `CARGO_TARGET_DIR=target` would still be available
+- We could add special behaviour like `CARGO_TARGET_BASE_DIR=""` meaning "use current" directory
+
+The first two are probably enough, the third is a bandaid.
+
+### Expose template metadata
+
+`cargo` will use backlinks in an implementation-defined form to keep track in the templated `CARGO_TARGET_DIR` of the relation from a target directory to its source workspace.
+
+In the future, we could envisage letting external tools and users access this data in a well-defined form through `cargo metadata`.
```

Comment:
This is a child under "Use templated `CARGO_TARGET_DIR` as the default instead of `target`" but not seeing how its related.

What metadata would we be exposing and why?  If we don't have an idea of what problem we're solving with this future possibility, it doesn't seem all that likely to happen

Reply:
> What metadata would we be exposing and why?

I have no idea, that's why it's a future possibilities. Someone *may* ask for metadata in the future but since I have no idea what would be needed, it's staying a vague future possibility with no plan to do it

Reply:
imo we should limit future possibilities to user-driven needs.

Reply:
Removed in https://github.com/rust-lang/rfcs/pull/3371/commits/8d1ddc29d8d637d0e2599d5aa31169704c30f511